File: javalin-openapi-processor/src/main/java/net/dzikoysk/openapi/processor/processing/AnnotationMirrorMapper.java
package net.dzikoysk.openapi.processor.processing;

import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.AnnotationValue;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.TypeMirror;
import java.util.List;
import java.util.Map.Entry;
import java.util.function.Function;
import java.util.function.IntFunction;

public class AnnotationMirrorMapper {

    protected final AnnotationMirror mirror;

    protected AnnotationMirrorMapper(AnnotationMirror mirror) {
        this.mirror = mirror;
    }

    protected Entry<? extends ExecutableElement, ? extends AnnotationValue> getEntry(String key) {
        return mirror.getElementValues().entrySet().stream()
                .filter(element -> element.getKey().getSimpleName().contentEquals(key))
                .findAny()
                .orElseThrow(() -> new IllegalStateException("Missing '" + key + "' property in @OpenApi annotation"));
    }

    protected Object getValue(String key) {
        return getEntry(key).getValue().getValue();
    }

    protected <R> R getAnnotation(String key, Function<AnnotationMirror, R> function) {
        return function.apply(getEntry(key).getValue().accept(new AnnotationVisitor<>(), null));
    }

    protected <T, R> R[] getArray(String key, Class<T> type, Function<T, R> mapper, IntFunction<R[]> arraySupplier) {
        return getArray(key, type).stream()
                .map(mapper)
                .toArray(arraySupplier);
    }

    protected <T> List<? extends T> getArray(String key, Class<T> type) {
        return getEntry(key).getValue().accept(new ArrayVisitor<T>(), null);
    }

    protected TypeMirror getType(String key) {
        return getEntry(key).getValue().accept(new ClassVisitor(), null);
    }

    protected String getString(String key) {
        return getValue(key).toString();
    }

    protected Boolean getBoolean(String key) {
        return Boolean.parseBoolean(getString(key));
    }

}

File: openapi-processor/src/main/kotlin/com/dzikoysk/openapi/processor/processing/AnnotationMirrorMapper.kt
package com.dzikoysk.openapi.processor.processing

import java.util.AbstractMap.SimpleEntry
import javax.lang.model.element.AnnotationMirror
import javax.lang.model.element.AnnotationValue
import javax.lang.model.element.ExecutableElement
import javax.lang.model.type.TypeMirror
import kotlin.collections.Map.Entry

internal open class AnnotationMirrorMapper protected constructor(protected val mirror: AnnotationMirror) {

    private fun getEntry(key: String): Entry<ExecutableElement, AnnotationValue> =
        mirror.elementValues.entries.firstOrNull { (element, _) -> element.simpleName.contentEquals(key) }
            ?: mirror.annotationType.asElement().enclosedElements.firstOrNull { it.simpleName.contentEquals(key) }
                ?.let {
                    val executableElement = it.accept(ExecutableVisitor(), null)
                    SimpleEntry(executableElement, executableElement.defaultValue)
                }
                ?: run { throw IllegalStateException("Missing '$key' property in @OpenApi annotation in ${mirror.annotationType.asElement().enclosingElement}") }

    protected fun getValue(key: String): Any =
        getEntry(key).value.value

    protected fun <R> getAnnotation(key: String, function: (AnnotationMirror) -> R): R =
        function(getEntry(key).value.accept(AnnotationVisitor(), null))

    protected fun <T, R> getArray(key: String, type: Class<T>, mapper: (T) -> R): List<R> =
        getArray(key, type).map(mapper)

    protected fun <T> getArray(key: String, @Suppress("UNUSED_PARAMETER") type: Class<T>): List<T> =
        getEntry(key).value.accept(ArrayVisitor(), null)

    protected fun getType(key: String): TypeMirror =
        getEntry(key).value.accept(ClassVisitor(), null)

    protected fun getString(key: String): String =
        getValue(key).toString()

    protected fun getBoolean(key: String) =
        getString(key).toBoolean()

}